File: BOJ/C++/1095.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

vector<int> p_list;
vector<int> p_count;

int EOSPC(int value, int prime)
{
  int res = 0;
  while(value > 1) {
    res += value / prime;
    value /= prime;
  }
  return res;
}

void find_all_prime(int s,int f,int m)
{
  for(int i=2;i<=m;i++)
  {
    bool prime = true;
    int limit = int(sqrt(i));
    for(int j=2;j<=limit;j++)
    {
      if(i % j == 0){
        prime = false;
        break;
      }
    }
    if(prime){
      int c = EOSPC(s+f,i) - EOSPC(s,i) - EOSPC(f,i);
        p_list.push_back(i);
        p_count.push_back(c);
    }
  }
}

int main() {
  int s,f,m;
  cin >> s >> f >> m;
  find_all_prime(s,f,m);
  int len = p_list.size();

  for(int i=m;i>1;i--)
  {
    bool ans = true;
    int target = i;
    for(int j = 0;j<len;j++)
    {
      int c = 0;
      int p = p_list[j];
      while(target % p == 0)
      {
        target /= p;
        c++;
        if(p_count[j] < c)
        {
          ans = false;
          break;
        }
      }
      if(!ans)
        break;
     }
    if(target == 1 && ans)
    {
      cout << i;
      return 0;
    }
  }
  cout << 1;
}

File: BOJ/Python/1095.py
import math
import time

def factorization(value,prime_table):
    res = []
    for p in prime_table:
        if value < p:
            break
        c = 0
        while value % p == 0:
            value //= p
            c += 1
        res.append(c)
    return res
    

def exp_of_single_prime_component(value,prime):
    res = 0
    while value > 1:
        res += value // prime
        value //= prime
    return res

def EOSPC(value,prime):
    return exp_of_single_prime_component(value,prime)

p_list = []
p_count = []
s,f,m = list(map(int,input().split()))

# Find all prime
for i in range(2,m+1):
    prime = True
    limit = int(math.sqrt(i))+1
    for j in range(2,limit):
        if i % j == 0:
            prime = False
            break
    if prime:
        p_list.append(i)
        c = EOSPC(s+f,i) - EOSPC(s,i) - EOSPC(f,i)
        p_count.append(c)
        if c > 0:
            res = i

start = time.time()
for i in range(m,res,-1):
    fac_i = factorization(i,p_list)
    l = len(fac_i)
    ans = True
    for j in range(l):
        if p_count[j] - fac_i[j] < 0:
            ans = False
            break
    if ans:
        res = i
        break
    
end = time.time()


print(res)

